core/src/price_estimation/orderbook_based.rs

```diff
@@ -0,0 +1,149 @@
+use super::price_source::PriceSource;
+use crate::models::{BatchId, TokenId};
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt as _};
+use pricegraph::{Orderbook, TokenPair};
+use primitive_types::U256;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+pub struct PricegraphEstimator {
+    orderbook_reader: Arc<dyn StableXOrderBookReading>,
```

Comment:
When wiring this into the threaded price source I wasn't able to get this to work with lifetimes (PriceOracle itself doesn't have any references so the only lifetime I could have used was static which the OrderbookReading isn't)

Reply:
I think we should be using `Arc` in most places instead of `&` anyway, it just makes things easier and the cost of arc is an atomic increment and decrement on `clone` and `drop`.